Add support for passing arguments to the test program

kjellwinblad opened the pull request:
This commit makes it possible to pass arguments to the test
program. The following example illustrate how this can be done:

$ nidhugg --sc test_program.ll argument1 argument2

margnus1 commented on src/Configuration.cpp:
```diff
@@ -24,6 +24,8 @@
 
 extern llvm::cl::opt<std::string> cl_transform;
 
+extern std::vector<std::string> cl_program_arguments;
```
I would move the `llvm::cl::list<std::string> arguments` from `main()` here. That way, you don't need the intermediary `vector<string>`.

I.e. you'd have `static llvm::cl::list<std::string> cl_program_arguments(`... here and no changes to `main.cpp`.

kjellwinblad replied:
The reason I did it this way is that cl_program_arguments(...) gets the argument llvm::cl::Positional so cl_program_arguments(...) needs to be called after input_file(...). If we do as I describe in the next answer below this will not be an issue anymore.

margnus1 replied:
I think @carlleonardsson will have to give the final word, but I would just add a `std::string input_file` member to `Configuration` and have all command line option objects be declared in `Configuration.cpp`.

kjellwinblad replied:
ok, I think I found another solution that will also fix the "--" problem. The solution is to use the [cl::ConsumeAfter modifier](https://llvm.org/docs/CommandLine.html#id22). I will try that out tomorrow.

margnus1 commented on src/Configuration.cpp:
```diff
@@ -112,6 +114,10 @@ void Configuration::assign_by_commandline(){
   print_progress = cl_print_progress || cl_print_progress_estimate;
   print_progress_estimate = cl_print_progress_estimate;
+  program_arguments.push_back("prog");
+  for(std::string a : cl_program_arguments){
```
You could use `vector.insert` here; but I admit that's a taste thing.

kjellwinblad replied:
i will change to vector.insert.

margnus1 commented on src/Configuration.h:
```diff
@@ -84,6 +85,7 @@ class Configuration{
     transform_loop_unroll = -1;
     print_progress = false;
     print_progress_estimate = false;
+    program_arguments = std::vector<std::string>();
```
There are two test suite failures in the debug (compile against a debug build of llvm, and pass `--enable-assertions` to `./configure`) build:
```
./unittest --report_level=short --log_level=warning --run_test='*'
Running 534 test cases...
unittest: /home/magnus/repos/llvm/lib/ExecutionEngine/ExecutionEngine.cpp:456: int llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, const std::vector<std::__cxx11::basic_string<char> >&, const char* const*): Assertion `!isTargetNullPtr(this, GVTOP(GVArgs[1])) && "argv[0] was null after CreateArgv"' failed.
unknown location(0): fatal error: in "POWER_test/Main_arg": signal: SIGABRT (application abort requested)
../../src/POWER_test2.cpp(887): last checkpoint: "Main_arg" test entry
unittest: /home/magnus/repos/llvm/lib/ExecutionEngine/ExecutionEngine.cpp:456: int llvm::ExecutionEngine::runFunctionAsMain(llvm::Function*, const std::vector<std::__cxx11::basic_string<char> >&, const char* const*): Assertion `!isTargetNullPtr(this, GVTOP(GVArgs[1])) && "argv[0] was null after CreateArgv"' failed.
unknown location(0): fatal error: in "POWER_test/Assertion_abort": signal: SIGABRT (application abort requested)
../../src/POWER_test2.cpp(903): last checkpoint: "Assertion_abort" test entry
```
What I think's happening is that since `Configuration::assign_by_commandline()` is not called when running the testsuite, `conf.program_arguments` stay the empty vector, and thus the test programs get `argc=0`, which is not a valid value for `argc`. Probably, if you just put `"prog"` in here instead, the failures will go away.

kjellwinblad replied:
Nice catch. I will implement the fix that you suggest.

kjellwinblad commented:
@margnus1 I updated the commit according to your suggestions.

* Unfortunately, [cl::ConsumeAfter modifier](https://llvm.org/docs/CommandLine.html#id22) did not work as stated by the documentation so one still has to pass "--" before the argument list but this has been documented now.
* I also changed so the name of the input ".ll"-file will be the first of the program's argument list when it is specified.
* For `nidhuggc`, I added an extra argument "--program_arguments" that can be used to specify arguments that will be passed to the test program. `nidhuggc` already makes use of "--" in its parameter list so correct parsing of the parameters would be quite difficult without introducing the special name "--program_arguments".
* @margnus1 Can you help me test if the unit tests run correctly now?

kjellwinblad commented:
Small fix

margnus1 commented:
I can confirm that the tests now pass.

I don't understand your argument about why it's hard to give `nidhuggc` the same syntax. As far as I understand it, you're concerned about the case `nidhuggc xx yy -- zz ww` being ambiguous, because it could either be clang and nidhugg arguments, or nidhugg and program arguments.
If that is you concern, then since the program file goes with the nidhugg arguments, and there are no clang flags that don't start with `-`, it seems to me that these ambigous commandlines can always be disambiguated based on whether there is anything among `xx` and `yy` which does not start with `-`. If so, we're always in the second case, and if not we're always in the first case.
It still seems to me that this could easily be implemented by rather than having `get_args()` always consider the last entry in `sys.argv` the input, you'd  instead set `A['--input']` if you encounter an argument that is not prefixed by `-` and `'--input'` is not already a key in `A`. Then, to make the handling of `--` correct, just add `'--input'` is not a key in `A` (`'--input' not in A`) as an additional condition for the `elif` clause that recognises `--`.

kjellwinblad commented:
@margnus1 you are right about my concern and I think your solution will work for all practical cases. Would it be ok if I drop my changes to nidhuggc.py so that you can add handling of arguments in nidhuggc.py after my commit?

margnus1 commented:
I don't mind that.

kjellwinblad commented:
@margnus1 You changes are added.

carlleonardsson commented:
Hi,

Is there a reason for the if-condition in Configure.cpp?

`if(cl_input_file.compare("-") != 0)`

It seems to me that the only effect of the condition is to get inconsistent behaviour for input files named "-". Or am I missing something?

On another note, I am a bit suspicious against having `argv[0]` be the name of the assembly file. I don't see how it would be used in a useful way. Especially considering that if nidhuggc is employed, that file name will be randomized on each invokation (so repeated runs of the same invokation of nidhuggc might yield different results).

kjellwinblad commented:
@carlleonardsson "-" is the default value for the input file. My reasoning behind this if statement is that if the input file is not set then it is better to keep the previous default value for `argv[0]`. However, I see know that perhaps it does not make sense if the input file always need to be set? Do you want be to change so it always use the previous default as `argv[0]`?

carlleonardsson commented:
@kjellwinblad yes, I think it is better to use a default program name for all tests independent of the assembly file name.

While you are at it, I suggest changing the default program name from `prog` to `a.out`, and to change the name of `Configuration::program_arguments` to `Configuration::argv` to make it more clear how it is supposed to be interpreted.

kjellwinblad commented:
@carlleonardsson I have done the changes that you suggested.
